File: data/scripts/upload_to_db.py
import os
from dotenv import load_dotenv
import pandas as pd
from pymongo import MongoClient

# 아래 기존 상대경로들 절대 경로로 변경해둔 것
BASE_DIR = os.path.dirname(os.path.abspath(__file__))
PROCESSED_DIR = os.path.join(BASE_DIR, "../processed")
parking_csv = os.path.join(PROCESSED_DIR, "parking_clean.csv")
ev_csv = os.path.join(PROCESSED_DIR, "ev_charger_cleaned.csv")

# --------------------------
# 환경변수 로드
# --------------------------
load_dotenv()
MONGO_URI = os.getenv("MONGO_URI")
print(MONGO_URI)

# --------------------------
# MongoDB 연결
# --------------------------
client = MongoClient(MONGO_URI)
db = client['seoul_smart_parking']  # 이미 만든 DB 이름

# --------------------------
# CSV 읽어서 컬렉션에 넣기
# --------------------------
# 1. 주차장

#절대경로로 변경함
#parking_csv = "C:/Users/hyoun/SeoulSmartParking/data/processed/parking_clean.csv"

parking_df = pd.read_csv(parking_csv)

# CSV -> dict list
parking_data = parking_df.to_dict(orient='records')

# 컬렉션에 삽입
db.parking.insert_many(parking_data)
print("✅ Parking 데이터 업로드 완료")

# 2. 전기차 충전소

#절대경로로 변경함
#ev_csv = "C:/Users/hyoun/SeoulSmartParking/data/processed/ev_charger_cleaned.csv"

ev_df = pd.read_csv(ev_csv)
ev_data = ev_df.to_dict(orient='records')
db.ev_charger.insert_many(ev_data)
print("✅ EV Charger 데이터 업로드 완료")

File: backend/utils/uploadToDb.js
// backend/utils/uploadToDb.js
import path from 'path';
import fs from 'fs';
import csv from 'csv-parser';
import { MongoClient } from 'mongodb';
import dotenv from 'dotenv';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// .env (backend/.env) 명시적으로 로드
dotenv.config({ path: path.join(__dirname, "../.env") });

// 프로젝트 루트 기준 data 폴더
const PROJECT_ROOT = path.join(__dirname, "..", "..");
const PROCESSED_DIR = path.join(PROJECT_ROOT, 'data', 'processed');
const PARKING_CSV = path.join(PROCESSED_DIR, 'parking_clean.csv');
const EV_CSV = path.join(PROCESSED_DIR, 'ev_charger_cleaned.csv');

// =========================
// MongoDB 연결
// =========================
const MONGO_URI = process.env.MONGO_URI;
if (!MONGO_URI) throw new Error('MONGO_URI 환경변수 설정 필요');

const client = new MongoClient(MONGO_URI);
const dbName = 'seoul_smart_parking';

// =========================
// CSV 읽기 함수
// =========================
async function readCsv(filePath) {
  return new Promise((resolve, reject) => {
    const results = [];
    if (!fs.existsSync(filePath)) return resolve([]);
    fs.createReadStream(filePath)
      .pipe(csv())
      .on('data', (data) => results.push(data))
      .on('end', () => resolve(results))
      .on('error', (err) => reject(err));
  });
}

// =========================
// DB 업로드 함수 (bulk upsert)
// =========================
export async function runUploadToDb() {
  try {
    await client.connect();
    console.log('✅ MongoDB 연결 완료');
    const db = client.db(dbName);

    // Parking: upsert by code (PKLT_CD -> code)
    const parkingData = await readCsv(PARKING_CSV);
    if (parkingData.length > 0) {
      const parkingOps = parkingData.map(doc => {
        // ensure key exists
        const filter = doc.code ? { code: doc.code } : { name: doc.name, address: doc.address };
        return {
          updateOne: {
            filter,
            update: { $set: doc },
            upsert: true
          }
        };
      });
      if (parkingOps.length) {
        await db.collection('parking').bulkWrite(parkingOps, { ordered: false });
        console.log('✅ Parking 데이터 업로드(업서트) 완료');
      }
    } else {
      console.log('ℹ️ parking_clean.csv 파일 없음 또는 비어있음');
    }

    // EV Charger: upsert by charging_station
    const evData = await readCsv(EV_CSV);
    if (evData.length > 0) {
      const evOps = evData.map(doc => {
        const filter = doc.charging_station ? { charging_station: doc.charging_station } : { address: doc.address };
        return {
          updateOne: {
            filter,
            update: { $set: doc },
            upsert: true
          }
        };
      });
      if (evOps.length) {
        await db.collection('ev_charger').bulkWrite(evOps, { ordered: false });
        console.log('✅ EV Charger 데이터 업로드(업서트) 완료');
      }
    } else {
      console.log('ℹ️ ev_charger_cleaned.csv 파일 없음 또는 비어있음');
    }

  } catch (err) {
    console.error('❌ DB 업로드 오류:', err);
  } finally {
    await client.close();
    console.log('MongoDB 연결 종료');
  }
}

// CLI 직접 실행 가능
if (process.argv[1] && process.argv[1].endsWith('uploadToDb.js')) {
  runUploadToDb();
}
